core/service/src/cryptography/signcryption.rs

```diff
@@ -10,27 +10,285 @@
+}
+
+pub trait DesigncryptFHEPlaintext: Designcrypt {
+    /// Decrypt a signcrypted plaintext message and verify the signature before returning the result.
+    /// If the signature verification fails, an error is returned.
+    /// The link parameter is used to verify that the signcryption corresponds to the expected context or session.
+    /// The method is exclusively used to decrypt partially decrypted FHE ciphertexts for user decryption.
+    fn designcrypt_plaintext(
+        &self,
+        dsep: &DomainSep,
+        signcryption: &[u8],
+        link: &[u8],
+    ) -> Result<SigncryptionPayload, CryptographyError>;
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Zeroize, VersionsDispatch)]
+pub enum UnifiedSigncryptionKeyOwnedVersioned {
+    V0(UnifiedSigncryptionKeyOwned),
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, Zeroize, Versionize)]
+#[versionize(UnifiedSigncryptionKeyOwnedVersioned)]
+pub struct UnifiedSigncryptionKeyOwned {
+    pub signing_key: PrivateSigKey,
+    pub receiver_enc_key: UnifiedPublicEncKey,
+    pub receiver_id: Vec<u8>, // Identifier for the receiver's encryption key, e.g. blockchain address
+}
+impl UnifiedSigncryptionKeyOwned {
+    pub fn new(
+        signing_key: PrivateSigKey,
+        receiver_enc_key: UnifiedPublicEncKey,
+        receiver_id: Vec<u8>,
+    ) -> Self {
+        Self {
+            signing_key,
+            receiver_enc_key,
+            receiver_id,
+        }
+    }
+
+    pub fn reference<'a>(&'a self) -> UnifiedSigncryptionKey<'a> {
+        UnifiedSigncryptionKey {
+            signing_key: &self.signing_key,
+            receiver_enc_key: &self.receiver_enc_key,
+            receiver_id: &self.receiver_id,
+        }
+    }
+}
+
+impl HasEncryptionScheme for UnifiedSigncryptionKeyOwned {
+    fn encryption_scheme_type(&self) -> EncryptionSchemeType {
+        self.receiver_enc_key.encryption_scheme_type()
+    }
+}
+impl HasSigningScheme for UnifiedSigncryptionKeyOwned {
+    fn signing_scheme_type(&self) -> SigningSchemeType {
+        self.signing_key.signing_scheme_type()
+    }
+}
+
+/// Internal type for signcryption keys, storing only references to the real internal keys.
+/// Thus this type should not be serialized instead `UnifiedSigncryptionKeyOwned` should be used.
+#[derive(Clone, Debug)]
+pub struct UnifiedSigncryptionKey<'a> {
+    pub signing_key: &'a PrivateSigKey,
+    pub receiver_enc_key: &'a UnifiedPublicEncKey,
+    pub receiver_id: &'a [u8], // Identifier for the receiver's encryption key, e.g. blockchain address
+}
+
+impl<'a> UnifiedSigncryptionKey<'a> {
+    pub fn new(
+        signing_key: &'a PrivateSigKey,
+        receiver_enc_key: &'a UnifiedPublicEncKey,
+        receiver_id: &'a [u8],
+    ) -> Self {
+        Self {
+            signing_key,
+            receiver_enc_key,
+            receiver_id,
+        }
+    }
+}
+
+impl HasEncryptionScheme for UnifiedSigncryptionKey<'_> {
```

Comment:
how useful is this? it seems `UnifiedSigncryptionKey` will always have an encryption scheme, so it can be a part of `UnifiedSigncryptionKey`

Reply:
I am not sure I understand the comment. You mean to remove the implementation at this level?
Sure, we can do that. It was just to avoid having to call on the inner part of a signcryption key to get the scheme